File: backend/test_api.js
const http = require('http');
const data = JSON.stringify({ subject: "cat-subj-cs" });
const options = { hostname: '127.0.0.1', port: 8080, path: '/catalog.v1.CatalogService/ListCourses', method: 'POST', headers: { 'Content-Type': 'application/json', 'Content-Length': data.length } };
const req = http.request(options, (res) => {
  let body = '';
  res.on('data', (chunk) => body += chunk);
  res.on('end', () => console.log('Response:', body));
});
req.on('error', (e) => console.error(e));
req.write(data);
req.end();

File: backend/test_api.py
import urllib.request
import json

def fetch_courses(search="", subject="", level="", sort_by=""):
    data = {
        "pageSize": 10,
        "searchQuery": search,
        "subject": subject,
        "level": level,
        "sortBy": sort_by
    }
    req = urllib.request.Request(
        "http://localhost:8000/catalog.v1.CatalogService/ListCourses",
        data=json.dumps(data).encode("utf-8"),
        headers={"Content-Type": "application/json"}
    )
    try:
        with urllib.request.urlopen(req) as response:
            res = json.loads(response.read().decode())
            print(f"Filters(search='{search}', subject='{subject}', level='{level}') -> {len(res.get('courses', []))} courses")
    except Exception as e:
        print(f"Error: {e}")

fetch_courses()
fetch_courses(search="Python")
fetch_courses(subject="cat-subj-ds")
